### DISCUSSION TITLE
Query optimization on primary key

### DISCUSSION THREAD

**Discussion by emilioschepis** [prisma/prisma#17523] [Q&A]
Hi everyone, I'm working on a project and I'm wondering what would be the fastest implementation for this scenario:

```prisma
model MyModel {
  id        String    @id @default(cuid())
  deletedAt DateTime?
}
```

If I want to find an instance by id but _only if it's not deleted_, which implementation is preferred?

**Option 1** (`findUnique`):
```ts
const model = await prisma.myModel.findUnique({
  where: {
    id: "<some-id>",
  },
  select: {
    deletedAt: true,
  }
})

if (model && model.deletedAt === null) {
  // Do something here
}
```

**Option 2** (`findFirst`):
```ts
const model = await prisma.myModel.findFirst({
  where: {
    AND: {
      id: "<some-id>",
      deletedAt: null
    }
  },
})

if (model) {
  // Do something here
}
```

**Comment by nurul3101** [comment]
Hello @emilioschepis 👋 

To get the best performance, I would recommend you create a combined index on the `id` and `deletedAt` fields like this:
```prisma
model MyModel {
  id        String    @id @default(cuid())
  deletedAt DateTime?

  @@id([id, deletedAt])
}
```

And then you can go with Option 2, On an additional note there is no need for `AND` in the query as by default all fields in where clause are combined with `AND`.
So this would be similar as the query using `AND`
```typescript
  const model = await prisma.myModel.findFirst({
    where: {
      id: "<some-id>",
      deletedAt: null,
    },
  });
```

Also, to validate this you can add [logging middleware](https://www.prisma.io/docs/concepts/components/prisma-client/middleware/logging-middleware) and check the query response time for both operations.

**Reply by emilioschepis** [reply]
Thank you @nurul3101, that's really helpful!

What if I had a more complex query, something that includes data from another table?
```ts
const model = await prisma.myModel.findFirst({
  where: {
    id: "<some-id>",
    deletedAt: null,
    users: {
      some: {
        userId: "<some-userid>",
        role: "owner",
      }
    }.
  },
});
```

I don't think I can make such a combined index. 
I guess I am trying to know if, with `id` being the primary key for `Model`, if the `findFirst` would stop as soon as it finds that element even if it does not match and return `null` or if it would still scan the whole table looking for another element with the same id (which would not be possible).

**Reply by nurul3101** [reply]
Yeah, I don't think you can create a combined index between different models.
The best way to determine how many documents were scanned by the database and whether database engine stopped or not as soon as id was matched would be by studying the query execution plan, you can use [explain analyze](https://www.postgresql.org/docs/current/sql-explain.html) to get the query execution plan.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
